### ISSUE TITLE
API Logging

### ISSUE THREAD

**Issue by Segelzwerg** [Whist-Team/Whist-Server#188]
Add more information on route access. E.g. 
`"POST /user/create HTTP/1.1" 200 OK` should contain the name of the user

**Comment by abant07** [comment]
What exactly is the problem

**Comment by Segelzwerg** [comment]
Hi @abant07 
I think we should extend the logging that it shows the names of users or games when they are created, logging in, or doing something else. For the moment we rely on the builtin logging of Uvicorn (I think) but should use our own logging in all API route methods.
Does this answer your question?

**Comment by abant07** [comment]
Hi @Segelzwerg 

Yes, thank you, the direction make sense, but I had a couple more questions

1. Is there an existing code that you created where I would need to extend upon or do I make brand new code? Also can I use Python?
2. I am not familiar with Uvicorn, can I know what exactly is logged into Uvicorn as you mentioned?

(This is my first open source contribution, so I'm sorry if things might go a little slow)

**Comment by Segelzwerg** [comment]
@abant07 thanks for your interest. I tried to answer your questions below, please feel free to ask again if something is not clear yet,

> 1. Is there an existing code that you created where I would need to extend upon or do I make brand new code? Also can I use Python?

No, until now we didn't log anything on our own and yes you actually should use Python as it is the only language we use for this repo. Maybe these links help: https://docs.python.org/3/howto/logging.html https://dev.to/tomas223/logging-tracing-in-python-fastapi-with-opencensus-a-azure-2jcm

> 2. I am not familiar with Uvicorn, can I know what exactly is logged into Uvicorn as you mentioned?

I guess it is somewhere documented, but in my opinion, we can write additional logs and don't care what Uvicorn does for the moment. So there might be some duplicate logs.

> (This is my first open source contribution, so I'm sorry if things might go a little slow)

Not to worry. Ask as much as you want and we work on this project in our free time only. So my understanding is you want to take care of this issue?

**Comment by abant07** [comment]
Yes, I would be happy to attempt this issue, however, I might frequently pop into this discussion chat if I have questions about the logging.

Thanks

**Comment by abant07** [comment]
Hi @Segelzwerg

So I just quickly wanted to pop in and clarify the assignment before I go too far. Do you want me to create a file that logs the name of the user when they are created and/or logged-in ? If that is the case, I'm kinda confused on how I would use an API route for this? Could you please advise me?

Thanks

**Comment by Segelzwerg** [comment]
Hi @abant07 there is no log them to a specific file it can be logged to the default logger, which in turn can later be logged to a file.

I hope this clears it up, otherwise please ask again.

**Comment by abant07** [comment]
so would the default logger just be the console? also am I right about logging the name of the user every time they log in/access the API route?

**Comment by Segelzwerg** [comment]
> so would the default logger just be the console?
Exactly. Only in the later stage we will create logging to file for production.

> also am I right about logging the name of the user every time they log in/access the API route?
Yes, unless you think it is not required for that route.

**Comment by abant07** [comment]
Thank you, so just to confirm. I am writing a function that is taking the input of the route access you have given (POST /user/create HTTP/1.1) and then parsing the user information from that route to then log it into the console?

Thanks,

**Comment by Segelzwerg** [comment]
> Thank you, so just to confirm. I am writing a function that is taking the input of the route access you have given (POST /user/create HTTP/1.1) and then parsing the user information from that route to then log it into the console?
> 
> Thanks,

Yes, that seems about right. You can also do a Draft PR if you want to discuss code.

**Comment by abant07** [comment]
Hi,

So I think I have got the code, but im not sure where to make the pull request.

**Comment by abant07** [comment]
Actually, nevermind, I pushed it to the server, let me know if you see it, otherwise, I can try it again. It's called "API Logging"

**Comment by Segelzwerg** [comment]
@abant07 I saw you forked the repo, but it is much easier to review for us when you create a PR. Can you do that?
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork

**Comment by abant07** [comment]
ok i have made pull request

**Comment by abant07** [comment]
Hi @Segelzwerg 

I have made a pull request a few days ago, but I'm not sure if the team is reviewing it. If you are not able to see it, I can try again.

Thanks

**Comment by iTitus** [comment]
see #198

**Comment by Segelzwerg** [comment]
@abant07 I do really appreciate your contribution, but I also agree with @iTitus that you may spend some time on how this application works.

Translating to this issue: The requirement for this is to add a little code snippet in each route and not write a new whole module. For an example for the `/game/join/{game_id}` you only need to add a log line within this code snippet:
```python
@router.post('/join/{game_id}', status_code=200)
def join_game(game_id: str, request: Dict[str, str], user: Player = Security(get_current_user),
              pwd_service=Depends(PasswordService), game_service=Depends(GameDatabaseService)):
    """
    User requests to join a game.
    :param game_id: unique identifier for a game
    :param request: must contain the key 'password'
    :param user: that tries to join the game. Must be authenticated.
    :param pwd_service: Injection of the password service. Required to create and check passwords.
    :param game_service: Injection of the game database service. Requires to interact with the
    database.
    :return: the status of the join request. 'joined' for successful join
    """


    game_pwd = request['password']
    game = game_service.get(game_id)


    if not pwd_service.verify(game_pwd, game.hashed_password):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Wrong game password.",
            headers={"WWW-Authenticate": "Basic"},
        )
    try:
        game.join(user)
        game_service.save(game)
    except PlayerAlreadyJoinedWarning:
        return {'status': 'already joined'}
    return {'status': 'joined'}

**Comment by abant07** [comment]
Thank you @Segelzwerg 

I think I am understanding after looking at the folders in the "code" section. So to clarify, is the issue that you want to add a log line to log the user game in each one of the files located in the "user" and "games" folder. For example, create.py, join.py, info.py, auth.py, etc.

Thanks

**Comment by Segelzwerg** [comment]
@abant07 Yes that is what were are looking for. You may need to have a look how the user object is constructed in order to get the username.

**Comment by abant07** [comment]
Hi @Segelzwerg 

I have just made the pull request. Please let me know if there is another problem needed to be fixed for this issue.

Thanks

**Comment by abant07** [comment]
Hi @Segelzwerg @iTitus 

Is everything good for this issue or is there still work to be done?

Thanks

**Comment by Segelzwerg** [comment]
@abant07 are you still working on this or can someone else take over?

**Comment by abant07** [comment]
Hi @Segelzwerg 

Someone can take over this.

Thanks
